### DISCUSSION TITLE
My Case Study Notes

### DISCUSSION THREAD

**Discussion by CamStLouis** [edwardkort/WWIDesigner#114] [Whistle Study]
Just wanted to share my ongoing case study notes (with questions for the developers highlighted). I hope this is helpful in demonstrating a practical workflow with the software. This case study centers around optimizing tonehole layouts for a new line pennywhistles I've developed.

The (read-only) link is here: https://docs.google.com/presentation/d/1sVuBJxuEatmckZ3mISAhhkvppwJreb-GtRasbT2f73A/edit?usp=sharing

**Comment by CamStLouis** [comment]
A number of questions have emerged, chief among them concerning the 21-inch whistle recommendation for a High Eb.

**Reply by CamStLouis** [reply]
For whatever reason, this is also the case now on the FeadogMk1 test files included in the original download. I have re-downloaded and overwritten the installation and that has not changed the result.

**Reply by bhp1** [reply]
From the console log, I see you are using the DIRECT global optimizer. Un-check the DIRECT checkbox in the options. For simple hole size and position optimization from a good starting point, you don't need global optimization. 

To be extra sure, on the constraints dialog reduce the upper bound on the bore length. 12" to 14" should be a reasonable value for an Eb whistle.

**Reply by CamStLouis** [reply]
Thanks! Could you clarify when one should and shouldn't use the DIRECT global optimizer?

**Reply by CamStLouis** [reply]
I am still encountering the problem where 4. Hole Size+Spacing optimizer just gives the largest possible hole size in the constraints file. Setting custom constraints just changed what size that was. (Skip to slide 24 in my notes)

**Reply by bhp1** [reply]
When should you use the global optimizer?  For now, stick with the local optimizer.  Resort to the global optimizer only if you are optimizing a number of interacting dimensions (like hole layout plus bore profile), and WIDesigner is missing a global optimum you are pretty sure is there somewhere, and you are prepared to experiment with different values of the constraints to guide DIRECT toward that global optimum.

As noted at [WIDesigner Optimizers](https://github.com/edwardkort/WWIDesigner/wiki/WIDesigner-Optimizers#optimization-algorithms), "the upper and lower bounds you use can make the difference between success and failure with DIRECT-C.}

**Reply by bhp1** [reply]
> I am still encountering the problem where 4. Hole Size+Spacing optimizer just gives the largest possible hole size in the constraints file.

Is that necessarily the wrong answer?  If you look at the tuning table or tuning graph, how well tuned are the notes over the two-octave range?  The real test would be to drill brass tubing with the suggested hole layout, stick a Feadog Mk 1 head on it, and see how well it plays in tune.

**Reply by CamStLouis** [reply]
Well, for a normal pennywhistle, it's definitely the wrong answer. Pennywhistles with massive holes like that will sound terrible in the second octave. It's not really an "optimizer" if it is generating hole layouts that don't serve the fundamental use case of the instrument. I tried with a specific min/max constraint for each hole consistent with the range of hole sizes measured on my existing instruments, but it still gave the biggest possible hole for each.

There is a reason that pretty much all pennywhistles have a similar pattern of hole spacing & sizing, and it's about balancing the two octaves as well as the tuning of the second octave by what is vented below the sounding hole.

Now, I have found that running the spacing optimizer alone, then the sizing optimizer, does seem to output a more reasonable suggested pattern, but optimizing separately does not really achieve the goal of finding an idealized layout.

**Reply by bhp1** [reply]
Unfortunately, WIDesigner has no measures of tone quality that it can consider in optimization.  At least for now, it is optimizing only for tuning.  From trying a variety of hole size maximum constraints, it looks like larger holes do give a marginal improvement in tuning over the full two octaves.  The improvement is so small, you couldn't notice it in practice, but the optimizer sees it, and that's all it has to go on.  At the moment, the only way to communicate "massive holes will sound terrible" to the optimizer is to reduce the maximum hole diameter in the constraints.  (Increasing the minimum hole spacing can also force the optimizer to use smaller holes.)  

For your high D whistle, maybe the best approach is to set the hole sizes according to what you think will give a good tone (and what drill bits you have available), and use the hole spacing optimizer alone.

> it's about balancing the two octaves as well as the tuning of the second octave

The optimizer does look at tuning in both octaves.  As for balancing the two octaves, for the example I tried, the supplementary information table suggests a whistle with smaller holes will take ever-so-slightly _more_ air in the second octave.

**Reply by bhp1** [reply]
BTW, can you elaborate on _how_ a whistle with big holes will sound terrible?  I can see that large holes would make it harder to play: getting clean notes at speed, and crisp cuts and taps.  Is there also an impact on tone quality?

**Reply by CamStLouis** [reply]
Ah, that makes sense. There is a function that predicts an ["optimal" bore for a given key,](https://www.music.bracker.uk/Music/Searching-For-The-Optimal-Whistle-Bore.html) and I find that selecting a bore slightly south of that gives my preferred tone, so I was hoping that it was something like that I could apply an offset factor to. No worries, that would be downright magical if it could! 

The reason why makers typically don't want to go with the largest holes the body will support is for two reasons. Number one is volume - it contributes to a piercing top hand in the second octave, although headjoint design also contributes to this. The other reason is that the size and placement of holes affects tone quality and response based on the acoustics of the individual note. This is the part I understand the least about woodwinds, and I'm working through a textbook right now to better understand it. Also, the pads of your fingers do occlude the bore to a greater degree with large holes and may affect tuning somewhat.

On flutes more than whistles, adjusting the two footjoint holes can tune the entire instrument. You can tune high G by the degree of undercut applied to the first footjoint hole. It's nuts! My understanding is that this is in part due to flutes being an exceptionally low-energy system, where the second-order acoustics of resonances and wave interactions have a much larger effect. I would desperately like to see a mathematical way to predict these interactions. 

Anyway, returning to whistles, these interactions are not as big a part of the tone, especially in a thinbody design. Some makers of thickbody whistles do undercut their holes, but I don't know if it's for the same reason. My priority, and why I'm using WwID, is to get consistency across the range of keys. I have absolutely _nailed_ Eb. It's everything a whistle should be, they come out consistent every time, and it's made 4 years of work worth it to have this design. However, the Ds I make have always been disappointing, either too "open" or too fuzzy. It wasn't until a recorder maker I was lucky enough to meet laid some downright _arcane_ shit on me that I was able to apply the scale factors correctly that gave me a D with the same alacrity and tone as the Eb. They're only a half step apart, but it might as well be a mile.

So, apart from minor tweaks to what my headjoint model predicts, the only thing left to solve for is the body tubes. If I can get WwID to output a quality hole configuration for my Eb (and I believe it has), I can use the same procedure and offset factors to get quality hole configurations for my other keys. I have never played a D whistle as good as an equivalent Eb until developing my own, and I'm eager to take that special tone and response to other keys.

I do hope that my questions about the tool don't seem overly critical; this is an incredible gift to instrument makers, and I really want to get the most it can do (and write some tutorials for others)! Optimizing an 18th century bagpipe is next, so I will have plenty more :sweat_smile:

**Comment by bhp1** [comment]
From the notes for slide 4:
> 2. Question: Should the bore points refer to the bore beginning at the edge of the airblade? I notice it is negative in the example by 0.2 of an inch (typical window length for a whistle of this type) - does that mean position is calculated at the edge of the airblade, but can be modeled taking into account dimensions that occur BEFORE the airblade (i.e. plug indentation, gap left by tuning slide etc)?

Yes, the bore can include points upstream of the blade.  The zero-origin of bore positions can be wherever is convenient.  For Gonzato-style low-tech whistles, the hole positions are fixed relative to the blade, but not relative to the fipple and the end of the windway, so I generally use the blade position as the origin.  For whistles, the "mouthpiece position" is the position of the blade.  With the blade as the origin, the end of the windway--the top of the bore--will have a negative position, and the mouthpiece position will be zero.  (In principle, you could use the bottom of the bore as the origin, so everything else would have a negative position.)

**Comment by bhp1** [comment]
From the notes for slide 13:
> Getting a better idea of what numerical values of deviation are expected would be helpful.
> Does the accuracy of the model hinge solely on the degree to which deviation in predicted and observed matches up?

The deviation columns are in cents.  I don't count on predictions to be better than 5 or 10 cents of measured values.  50 cents, however, is half a semitone: a big miss.

(When using min/max frequencies to calibrate the model, the calibrator aims to minimize those deviations.  The deviations between target frequency and predicted playing frequency don't come into play.)

**Comment by bhp1** [comment]
Again, from the notes for slide 4:
> 2b  The bore of this whistle can best be described as a “stepped cylinder;” assuming that WwID assumes a linear transition between bore point, this necessitated marking a “start” and “end” position for each cylinder. Assuming WwID cannot accept two diameters with the same position, this necessitated fudging a bit, starting the next cylinder a thou away from the previous.

Yes, WIDesigner assumes a linear taper between bore points.  For a stepped cylinder, a pair of bore points a short distance apart is the appropriate way to model the step.  I'd suggest more like 5-10 thou apart.

**Comment by Sonoton** [comment]
I would like to view this material.  Can you provide access for me for the Google Drive?
Sonoton
